snowpack/src/build/optimize.ts

```diff
@@ -0,0 +1,538 @@
+      scriptRoot.removeAttr('snowpack-inline');
+      unlinkSync(scriptFile);
+    });
+  getLinks('stylesheet')
+    .filter('[snowpack-inline]')
+    .each((_, elem) => {
+      const linkRoot = root(elem);
+      const styleFile = path.resolve(file, '..', linkRoot.attr('href')!);
+      const styleContent = readFileSync(styleFile, 'utf-8');
+      const newStyleEl = root('<style></style>');
+      newStyleEl.text(styleContent);
+      linkRoot.after(newStyleEl);
+      linkRoot.remove();
+      unlinkSync(styleFile);
+    });
+}
+
+/** Add new bundled CSS files to the HTML entrypoint file, if not already there. */
+function addNewBundledCss(
+  htmlData: ScannedHtmlEntrypoint,
+  manifest: ESBuildMetaManifest,
+  baseUrl: string,
+): void {
+  if (!manifest.outputs) {
+    return;
+  }
+  for (const key of Object.keys(manifest.outputs)) {
+    if (!key.endsWith('.css')) {
+      continue;
+    }
+    const scriptKey = key.replace('.css', '.js');
+    if (!manifest.outputs[scriptKey]) {
+      continue;
+    }
+    const hasCssImportAlready = htmlData
+      .getLinks('stylesheet')
+      .toArray()
+      .some((v) => v.attribs.href.includes(removeLeadingSlash(key)));
+    const hasScriptImportAlready = htmlData
+      .getScripts()
+      .toArray()
+      .some((v) => v.attribs.src.includes(removeLeadingSlash(scriptKey)));
+
+    if (hasCssImportAlready || !hasScriptImportAlready) {
+      continue;
+    }
+    const linkHref = removeTrailingSlash(baseUrl) + addLeadingSlash(key);
+    htmlData.root('head').append(`<link rel="stylesheet" href="${linkHref}" />`);
+  }
+}
+
+/**
+ * Traverse the entrypoint for JS scripts, and add preload links to the HTML entrypoint.
+ */
+function preloadEntrypoint(
+  htmlData: ScannedHtmlEntrypoint,
+  manifest: ESBuildMetaManifest,
+  config: SnowpackConfig,
+): void {
+  const {root, getScripts} = htmlData;
+  const preloadScripts = getScripts()
+    .map((_, elem) => elem.attribs.src)
+    .get()
+    .filter(isTruthy);
+  const collectedDeepImports = new Set<string>();
+  for (const preloadScript of preloadScripts) {
+    const normalizedPreloadScript = removeLeadingSlash(
+      path.posix.resolve('/', removeLeadingSlash(preloadScript)),
+    );
+    collectDeepImports(normalizedPreloadScript, manifest, collectedDeepImports);
+  }
+  const baseUrl = config.buildOptions.baseUrl;
+  for (const imp of collectedDeepImports) {
+    const preloadUrl = (baseUrl ? removeTrailingSlash(baseUrl) : '') + addLeadingSlash(imp);
+    root('head').append(`<link rel="modulepreload" href="${preloadUrl}" />`);
+  }
+}
+
+/**
+ * Given any set of user-input as entrypoints,
+ */
+async function resolveEntrypoints(
+  entrypoints: OptimizeOptions['entrypoints'],
+  allBuildFiles: string[],
+) {
+  if (entrypoints === 'auto') {
+    // TODO: Filter allBuildFiles by HTML with head & body
+    return allBuildFiles.filter((f) => f.endsWith('.html'));
+  }
+  if (Array.isArray(entrypoints)) {
+    return entrypoints;
```

Comment:
WDYT about something like this?

```diff
  if (Array.isArray(entrypoints)) {
-    return entrypoints;
+    return entrypoints.map(file => file.startsWith(path.sep) ? file : path.join(cwd, file))
```

Basically, use an absolute path if provided by the user, or assume it‘s relative to the project directory (see above comment for more context)

Reply:
Done, added handling for relative paths to support both final build directory and source directory.